server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java

```diff
@@ -62,14 +74,98 @@ public void uploadBlobAsync(
     }
 
     @Override
-    public void uploadBlob(final TransferFileSnapshot fileSnapshot, Iterable<String> remoteTransferPath) throws IOException {
-        assert remoteTransferPath instanceof BlobPath;
+    public void uploadBlob(final TransferFileSnapshot fileSnapshot, Iterable<String> remoteTransferPath, WritePriority writePriority)
+        throws IOException {
         BlobPath blobPath = (BlobPath) remoteTransferPath;
         try (InputStream inputStream = fileSnapshot.inputStream()) {
             blobStore.blobContainer(blobPath).writeBlobAtomic(fileSnapshot.getName(), inputStream, fileSnapshot.getContentLength(), true);
         }
     }
 
+    @Override
+    public void uploadBlobs(
+        Set<TransferFileSnapshot> fileSnapshots,
+        final Map<Long, BlobPath> blobPaths,
+        ActionListener<TransferFileSnapshot> listener,
+        WritePriority writePriority
+    ) {
+        List<CompletableFuture<Void>> resultFutures = new ArrayList<>();
+        fileSnapshots.forEach(fileSnapshot -> {
+            BlobPath blobPath = blobPaths.get(fileSnapshot.getPrimaryTerm());
+            if (!(blobStore.blobContainer(blobPath) instanceof VerifyingMultiStreamBlobContainer)) {
+                uploadBlobByThreadPool(ThreadPool.Names.TRANSLOG_TRANSFER, fileSnapshot, blobPath, listener, writePriority);
+            } else {
+                CompletableFuture<Void> resultFuture = createUploadFuture(fileSnapshot, listener, blobPath, writePriority);
+                if (resultFuture != null) {
+                    resultFutures.add(resultFuture);
+                }
+            }
+        });
+
+        if (resultFutures.isEmpty() == false) {
+            CompletableFuture<Void> resultFuture = CompletableFuture.allOf(resultFutures.toArray(new CompletableFuture[0]));
+            try {
+                resultFuture.get();
+            } catch (Exception e) {
+                logger.warn("Failed to upload blobs", e);
+            }
+        }
```

Comment:
Might be better to reuse `uploadBlobByThreadPool` in a loop and minimise duplication? Also it might be great if we were blocking on the response at `AsyncIOProcessor` rather than at multiple places

Reply:
Threadpool method is using sync api for upload. That cannot be used. createUploadFuture method is doing exactly what you are saying.

Reply:
Blocking in asyncIOProcessor will require decent amount of changes. I created a github earlier for this. https://github.com/opensearch-project/OpenSearch/issues/8108